### DISCUSSION TITLE
Welcome to OOHTML Discussions!

### DISCUSSION THREAD

**Discussion by ox-harris** [webqit/oohtml#1] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by OliverBrotchie** [comment]
Hey @ox-harris, I am going to be writing a compiler for OOHTML in rust. I was wondering, is there a way to take HTML as input to a template?? Eg:
```html
<head>
    <template name="title-module">
        <export name="title>
            <nav>
                 <arg name="foo"/>
            </nav>
        </export>
    </template>
</head>
<body>
    <import name="title" template="title-module">
        <arg name="foo">
            <h1>This is a title!!<h1>
        </arg>
    </import>
</body>
```
Which would output as:
```html
<body>
    <nav>
         <h1>This is a title!!</h1>
    </nav>
</body>
```

This would be a huge improvement for noScript users!

**Reply by ox-harris** [reply]
I am definitely interested in your compiler idea.

**Reply by OliverBrotchie** [reply]
@ox-harris Good to hear it! I have created a basic repository over at [OliverBrotchie/mosaic](https://github.com/OliverBrotchie/mosaic).

**Reply by ox-harris** [reply]
Great I'll be following.

**Reply by ox-harris** [reply]
Hi @OliverBrotchie , I hope it's all coming up well.

I wanted to share that we now have an official page for recommending tooling for OOHTML.  Page: https://webqit.io/tooling/oohtml/docs/resources/ecosystem. I'll be more than glad to add a rust compiler for OOHTML.

**Reply by OliverBrotchie** [reply]
Dear @ox-harris, apologies for not getting back to you! I have been quite busy with other projects and haven't had time to work on this yet, but I will let you know if I make any progress.

**Reply by ox-harris** [reply]
No problem @OliverBrotchie. Anticipating when you will be freer.

**Comment by ox-harris** [comment]
Indeed! And thanks for the good thinking!

OOHTML would let you do this. And here is how it goes (from OOHTML v1.7.5):

1. **To receive an element as an arg...**
   
     Instead of using some `<arg />` element as you have in the code:
    
    ```html
    <head>

        <template name="title-module">
            <export name="title">
                <nav>
                     <arg name="foo"/>
                </nav>
            </export>
        </template>

    </head>
    ```
   
     we would use a regular `<import></import>` element:
    
    ```html
    <head>

        <template name="title-module">
            <export name="title">
                <nav>
                     <import name="foo"></import>
                </nav>
            </export>
        </template>

    </head>
    ```
   
     The idea is that the host `<export></export>` element will automatically inherit content from whatever the slot it is eventually resolved into in the main document, in this case, the *title* slot. Now, the `<import></import>` element would be scoped to the host `<export></export>` element to import the inherited content.

    We use the *tilde* character "~" to resolve imports from their host `<export></export>` element:
    
    ```html
    <head>

        <template name="title-module">
            <export name="title">
                <nav>
                     <import name="foo" template="~"></import>
                </nav>
            </export>
        </template>

    </head>
    ```

    The last part is to properly construct the content that is to be inherited at resolve-time...

2. **To put content up for inheritance...**

     Instead of using some `<arg />` element as you have in the code:
    
    ```html
    <body>

        <import name="title" template="title-module">
            <arg name="foo">
                <h1>This is a title!!<h1>
            </arg>
        </import>

    </body>
    ```

    we would use a regular `<export></export>` element:
    
    ```html
    <body>

        <import name="title" template="title-module">
            <export name="foo">
                <h1>This is a title!!<h1>
            </export>
        </import>

    </body>
    ```

Hope that looks good!

It might also be good to see  how this looks in the alternative *import/export* syntax that uses the `exportgroup` attribute instead of an `<export></export>` element:
  
1. **a**

    ```html
    <head>

        <template name="title-module">
            <nav exportgroup="title">
                 <import name="foo" template="~"></import>
            </nav>
        </template>

    </head>
    ```

2. **b**

    ```html
    <body>

        <import name="title" template="title-module">
            <h1 exportgroup="foo">This is a title!!<h1>
        </import>

    </body>
    ```

And that's it! Fewer elements up there!

And talking about less code, the idea of *default import/export* in OOHTML could further simplify things as it lets us do without import and export IDs.

Below we omit the `foo` import and export ID:
  
1. **a**

    ```html
    <head>

        <template name="title-module">
            <nav exportgroup="title">
                 <import template="~"></import>
            </nav>
        </template>

    </head>
    ```

2. **b**

    ```html
    <body>

        <import name="title" template="title-module">
            <h1>This is a title!!<h1>
        </import>

    </body>
    ```

Then, if we were to have to structure the inherited content in some special way, we could use the template element:
  
1. **a**

    ```html
    <head>

        <template name="title-module">
            <nav exportgroup="title">
                 <import template="~/nested"></import>
            </nav>
        </template>

    </head>
    ```

2. **b**

    ```html
    <body>

        <import name="title" template="title-module">
            <h1>This is a title!!<h1>
            <template name="nested">
                <h2>This is nested title</h2>
            </template>
        </import>

    </body>
    ```

    > Note that`<template></template>` elements within `<import></import>` elements do not appear in the global module namespace.

**Reply by OliverBrotchie** [reply]
@ox-harris Oh wow that's a perfect solution! My only worry is that it might not be entirely intuitive behaviour to developers. But I'm sure with some good documentation on the logic behind import/export with examples, people will understand it well enough.

**Reply by ox-harris** [reply]
You are right @OliverBrotchie . We're currently working to come up with better documentation.

**Comment by OliverBrotchie** [comment]
@ox-harris I have another few of questions. 

## Question One

How are CSS and JS imports handled? Are they imported inline or is the CSS moved to the head and the JS appended to the body?

### Example

Would this:

```html
<!-- template.html -->
<export name="title>
    <style type="text/css">
        h1 {
            color: white;
        }
    </style>
    <h1>This is a title!!</h1>
    <script> console.log("Title module has been imported")</script>
</export>
```

```html
<!-- import.html -->
<head>
    <template name="title-module" src="/template.html"/>
</head>
<body>
    <import template="title-module" name="title"/>
    <import template="title-module" name="title"/>
</body>
```

Compile to this:

```html
<head>
    <style type="text/css">
        h1 {
            color: white;
        }
    </style>
</head>
<body>
    <h1>This is a title!!</h1>
    <h1>This is a title!!</h1>
    
    <script> console.log("Title module has been imported")</script>
    <script> console.log("Title module has been imported")</script>
</body>
```

Or this:
```html
<body>
    <style type="text/css">
        h1 {
            color: white;
        }
    </style>
    <h1>This is a title!!</h1>
    <script> console.log("Title module has been imported")</script>
    <style type="text/css">
        h1 {
            color: white;
        }
    </style>
    <h1>This is a title!!</h1>
    <script> console.log("Title module has been imported")</script>
</body>
```

## Question Two 

What is the correct way to import HTML templates from rel?

Should they just be a HTML snippet like this?

```html
<export name="title>
    <style type="text/css">
        h1 {
            color: white;
        }
    </style>
    <h1>This is a title!!</h1>
    <script> console.log("Title module has been imported")</script>
</export>
```

Or should they be a whole document like this?

```html
<!DOCTYPE html>
<html lang="en">
<head>
    <export name="title>
        <style type="text/css">
            h1 {
                color: white;
            }
        </style>
        <h1>This is a title!!</h1>
        <script> console.log("Title module has been imported")</script>
    </export>
</head>
```

If you are just supposed to use the former, how should the browser handle tags such as head, body etc in the imports?

## Question Three

Would it not be more prudent to use href attribute instead of rel for linking to templates? This would allow the importing of templates cross-origin.

Eg:

```html
 <template name="title-module" href="http://website/some/url/template.html"/>
 ```
 
 I believe the rel tag usually defines the type of relationship between a linked resource, not the location of the resource.
 Perhaps it could be used to answer the previous question with: 

 ```html
 <!-- rel="snippet" for non-well-formed HTML documents -->
 <template name="title-module" href="http://website/some/url/template.html" rel="snippet" />
 
 <!-- rel="document" for importing well-formed HTML documents -->
 <template name="title-module" href="http://website/some/url/template.html" rel="document" />
 ```

**Reply by ox-harris** [reply]
For **question one**, the latter is the case. `<import />` elements are simply replaced by the referenced *exports*.

For **question two**, the former is the case. Remote contents are not expected to be fully formed HTML documents. Remote contents with `<head />` and `<body />` tags would obviously result in a malformed document when placed with the `<import />` element. But programatically handling them by accessing the `document.templates` object would work just fine:

 ```html
<head>
    <template name="fullDoc" src="/document.html"></template>
    <script>
        let fullDoc = document.templates.fullDoc;
        // Being a remote module, we'd first ensure that content is loaded before accessing them
        fullDoc.addEventListener('load', e => {
            // If we were to access content via the template.content property
            fullDoc.content.children;
        
            // If we were to access exports via the template.exports property
            fullDoc.exports;
        });
    </script>
</head>
```

For **question three**, did you mean to say **rel** or **src**? What OOHTML uses is actually **src**.

**Reply by OliverBrotchie** [reply]
@ox-harris These answers defiantly make sense, and you are correct, I misread the document on question three haha.

I guess for question one, it would be more performant in most cases to import all the CSS in the head, but that should be left up to the developer. 

Thanks for your quick response!

**Reply by ox-harris** [reply]
@OliverBrotchie I am more than glad at the conversation! Many thanks.

**Comment by ox-harris** [comment]
Pleased to announce our new OOHTML docs: https://webqit.io/tooling/oohtml

**Reply by boticello** [reply]
This is a beautiful sharpening of HTML. All your design decisions are elegant. I will enjoy using it.

**Reply by ox-harris** [reply]
Super stoked to hear that @boticello
